libraries/AP_RangeFinder/AP_RangeFinder_JRE_Serial.cpp

```diff
@@ -0,0 +1,112 @@
+{
+    // uart instance check
+    if (uart == nullptr) {
+        return false;  // not update
+    }
+    uint32_t n = uart->available();
+    if (n == 0) {
+        return false;
+    }
+
+    uint16_t count = 0;
+    float reading_cm;
+    // max distance the sensor can reliably measure - read from parameters
+    const int16_t distance_cm_max = max_distance_cm();
+
+    // buffer read
+    const ssize_t num_read = uart->read(read_buff, ARRAY_SIZE(read_buff));
+    read_buff_idx = 0; // read_buff start data index
+    while (read_buff_idx < num_read) {
+
+        if (data_buff_idx == 0) { // header first byte check
+            // header data search
+            for (; read_buff_idx<num_read; read_buff_idx++) {
+                if (read_buff[read_buff_idx] == FRAME_HEADER_1) {
+                    data_buff[0] = FRAME_HEADER_1;
+                    data_buff_idx = 1; // next data_buff
+                    if (read_buff_idx >= num_read - 1) { // read end byte
+                        if (count > 0) {
+                            return true;
+                        }
+                        return false;      // next packet to header second byte
+                    } else {
+                        if (read_buff[read_buff_idx + 1] == FRAME_HEADER_2) {
+                            data_buff[1] = FRAME_HEADER_2;
+                            data_buff_idx = 2;    // next data_buff
+                            read_buff_idx += 2;   // next read_buff
+                            break;         // next data set
+                        } else {
+                            data_buff_idx = 0;    // data index clear
+                        }
+                    }
+                }
+            }
+
+        } else if (data_buff_idx == 1) { // header second byte check
+            if (read_buff[read_buff_idx] == FRAME_HEADER_2) {
+                data_buff[1] = FRAME_HEADER_2;
+                data_buff_idx = 2;    // next data_buff
+            } else {
+                data_buff_idx = 0;    // data index clear
+            }
+            read_buff_idx++;      // next read_buff
+
+        } else { // data set
+            if (data_buff_idx >= ARRAY_SIZE(data_buff)) {  // 1 data set complete
+                // crc check
+                uint16_t crc = crc16_ccitt_r(data_buff, ARRAY_SIZE(data_buff) - 2, 0xffff, 0xffff);
+                if ((((crc>>8) & 0xff) == data_buff[15]) && ((crc & 0xff) == data_buff[14])) {
+                    // status check
+                    if (data_buff[13] & 0x02) { // NTRK
+                        no_signal = true;
+                        reading_m = MIN(MAX(DIST_MAX_CM, distance_cm_max), UINT16_MAX) * 0.01f;
+                    } else { // UPDATE DATA
+                        no_signal = false;
+                        reading_cm = data_buff[4] * 256 + data_buff[5];
+                        if (reading_cm < distance_cm_max) {
+                            reading_m = reading_cm * 0.01f;
```

Comment:
It looks like this function will only ever return the last distance that it is in the buffer at the time this get_reading() is run.  Normally our rangefinder drivers return the average of all readings.
I think this also means that if it getse two readings and the last one is "NTRK" it will return no_signal instead of the earlier good reading.